File: backend/src/main/java/com/crashedcarsales/entity/Car.java
package com.crashedcarsales.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "cars")
public class Car {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "dealer_id", nullable = false)
    private Dealer dealer;

    @NotBlank(message = "Make is mandatory")
    @Column(name = "make", nullable = false)
    private String make;

    @NotBlank(message = "Model is mandatory")
    @Column(name = "model", nullable = false)
    private String model;

    @NotNull(message = "Year is mandatory")
    @Min(value = 1886, message = "Year must be valid")
    @Column(name = "year", nullable = false)
    private Integer year;

    @NotNull(message = "Fuel type is mandatory")
    @Enumerated(EnumType.STRING)
    @Column(name = "fuel_type", nullable = false)
    private FuelType fuelType;

    @NotNull(message = "Transmission is mandatory")
    @Enumerated(EnumType.STRING)
    @Column(name = "transmission", nullable = false)
    private Transmission transmission;

    @NotNull(message = "Vehicle type is mandatory")
    @Enumerated(EnumType.STRING)
    @Column(name = "vehicle_type", nullable = false)
    private VehicleType vehicleType;

    @NotNull(message = "Condition is mandatory")
    @Enumerated(EnumType.STRING)
    @Column(name = "condition", nullable = false)
    private Condition condition;

    @NotNull(message = "Price is mandatory")
    @DecimalMin(value = "0.0", inclusive = false, message = "Price must be greater than 0")
    @Column(name = "price", nullable = false)
    private BigDecimal price;

    @Min(value = 0, message = "Mileage cannot be negative")
    @Column(name = "mileage")
    private Integer mileage;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "is_featured")
    private Boolean isFeatured = false;

    @Column(name = "is_active")
    private Boolean isActive = true;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "car", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<CarImage> images = new ArrayList<>();

    @OneToOne(mappedBy = "car", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private CarFeatures features;

    // Constructors
    public Car() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        this.isFeatured = false;
        this.isActive = true;
    }

    public Car(Dealer dealer, String make, String model, Integer year, FuelType fuelType,
               Transmission transmission, VehicleType vehicleType, Condition condition, BigDecimal price) {
        this();
        this.dealer = dealer;
        this.make = make;
        this.model = model;
        this.year = year;
        this.fuelType = fuelType;
        this.transmission = transmission;
        this.vehicleType = vehicleType;
        this.condition = condition;
        this.price = price;
    }

    // Lifecycle callbacks
    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    // Getters and Setters
    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public Dealer getDealer() {
        return dealer;
    }

    public void setDealer(Dealer dealer) {
        this.dealer = dealer;
    }

    public String getMake() {
        return make;
    }

    public void setMake(String make) {
        this.make = make;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public Integer getYear() {
        return year;
    }

    public void setYear(Integer year) {
        this.year = year;
    }

    public FuelType getFuelType() {
        return fuelType;
    }

    public void setFuelType(FuelType fuelType) {
        this.fuelType = fuelType;
    }

    public Transmission getTransmission() {
        return transmission;
    }

    public void setTransmission(Transmission transmission) {
        this.transmission = transmission;
    }

    public VehicleType getVehicleType() {
        return vehicleType;
    }

    public void setVehicleType(VehicleType vehicleType) {
        this.vehicleType = vehicleType;
    }

    public Condition getCondition() {
        return condition;
    }

    public void setCondition(Condition condition) {
        this.condition = condition;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public Integer getMileage() {
        return mileage;
    }

    public void setMileage(Integer mileage) {
        this.mileage = mileage;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Boolean getIsFeatured() {
        return isFeatured;
    }

    public void setIsFeatured(Boolean isFeatured) {
        this.isFeatured = isFeatured;
    }

    public Boolean getIsActive() {
        return isActive;
    }

    public void setIsActive(Boolean isActive) {
        this.isActive = isActive;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public List<CarImage> getImages() {
        return images;
    }

    public void setImages(List<CarImage> images) {
        this.images = images;
    }

    public CarFeatures getFeatures() {
        return features;
    }

    public void setFeatures(CarFeatures features) {
        this.features = features;
    }

    // Enums
    public enum FuelType {
        GAS, HYBRID, DIESEL, ELECTRIC
    }

    public enum Transmission {
        MANUAL, AUTOMATIC, CVT
    }

    public enum VehicleType {
        VAN, MOTOR, PASSENGER, TRUCK
    }

    public enum Condition {
        DAMAGED, USED, ACCIDENTED, DERELICT
    }
}

File: frontend/src/types/car.ts
export interface CarImage {
  id: string;
  carId: string;
  imageUrl: string;
  altText?: string;
  displayOrder: number;
  createdAt: string;
}

export interface CarFeatures {
  airbags: boolean;
  absBrakes: boolean;
  airConditioning: boolean;
  powerSteering: boolean;
  centralLocking: boolean;
  electricWindows: boolean;
}

export interface Car {
  id: string;
  dealerId: string;
  make: string;
  model: string;
  year: number;
  fuelType: 'GAS' | 'HYBRID' | 'DIESEL' | 'ELECTRIC';
  transmission: 'MANUAL' | 'AUTOMATIC' | 'CVT';
  vehicleType: 'VAN' | 'MOTOR' | 'PASSENGER' | 'TRUCK';
  condition: 'DAMAGED' | 'USED' | 'ACCIDENTED' | 'DERELICT';
  price: number;
  mileage?: number;
  description?: string;
  images: CarImage[];
  features: CarFeatures;
  isFeatured: boolean;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  dealer?: Dealer;
}

export interface Dealer {
  id: string;
  name: string;
  email: string;
  address?: string;
  phone?: string;
  website?: string;
}

export interface CarSearchRequest {
  query?: string;
  make?: string;
  model?: string;
  fuelType?: Car['fuelType'];
  transmission?: Car['transmission'];
  vehicleType?: Car['vehicleType'];
  condition?: Car['condition'];
  minPrice?: number;
  maxPrice?: number;
  minYear?: number;
  maxYear?: number;
  maxMileage?: number;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
}

export interface PaginatedResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
  numberOfElements: number;
}

export interface CarFilters {
  makes: string[];
  models: string[];
  fuelTypes: Car['fuelType'][];
  transmissions: Car['transmission'][];
  vehicleTypes: Car['vehicleType'][];
  conditions: Car['condition'][];
  priceRange: {
    min: number;
    max: number;
  };
  yearRange: {
    min: number;
    max: number;
  };
}

export interface SearchState {
  query: string;
  filters: CarSearchRequest;
  results: Car[];
  totalResults: number;
  currentPage: number;
  totalPages: number;
  isLoading: boolean;
  error: string | null;
  viewMode: 'grid' | 'list';
}